File: server/_core/cognito.ts
/**
 * AWS Cognito認証
 * 
 * AWS Cognitoを使用したJWTベースの認証機能を提供する。
 * Bearerトークン方式でリクエストを認証し、ユーザー情報を取得する。
 * 
 * 認証フロー:
 * 1. リクエストからBearerトークンを取得
 * 2. CognitoのJWKSエンドポイントから公開鍵を取得
 * 3. JWTを検証（署名、発行者、有効期限など）
 * 4. ユーザー情報をデータベースに保存/更新
 * 5. ユーザーオブジェクトを返す
 */
import { ForbiddenError } from "@shared/_core/errors";
import { createRemoteJWKSet, jwtVerify } from "jose";
import type { Request } from "express";
import type { User } from "../../drizzle/schema";
import * as db from "../db";
import { ENV } from "./env";

/**
 * Cognitoの発行者URLを取得する
 * 
 * CognitoのユーザープールIDとリージョンから発行者URLを生成する。
 * 環境変数が設定されていない場合は空文字列を返す。
 * 
 * @returns Cognitoの発行者URL、または空文字列（未設定の場合）
 */
const getIssuer = () => {
  if (!ENV.cognitoRegion || !ENV.cognitoUserPoolId) return "";
  return `https://cognito-idp.${ENV.cognitoRegion}.amazonaws.com/${ENV.cognitoUserPoolId}`;
};

/**
 * CognitoのJWKS URL
 * 
 * 環境変数で設定されている場合はそれを使用し、
 * 未設定の場合は発行者URLから自動生成する。
 * JWKS（JSON Web Key Set）はJWTの検証に使用する公開鍵のセット。
 */
const jwksUrl =
  ENV.cognitoJwksUrl ||
  (ENV.cognitoRegion && ENV.cognitoUserPoolId
    ? `${getIssuer()}/.well-known/jwks.json`
    : "");

/**
 * CognitoのJWKS（リモート公開鍵セット）
 * 
 * JWTの署名を検証するために使用する公開鍵セット。
 * リモートエンドポイントから取得し、キャッシュされる。
 */
const jwks = jwksUrl ? createRemoteJWKSet(new URL(jwksUrl)) : null;

/**
 * Bearerトークンを取得する
 * 
 * リクエストのAuthorizationヘッダーからBearerトークンを抽出する。
 * Bearerトークンは"Bearer <token>"の形式で送信される。
 * 
 * 処理の流れ:
 * 1. Authorizationヘッダーを取得（大文字小文字を区別しない）
 * 2. ヘッダー値を" "で分割してトークンタイプとトークンを取得
 * 3. トークンタイプが"bearer"の場合のみトークンを返す
 * 
 * @param req - Expressリクエストオブジェクト
 * @returns Bearerトークン、またはnull（トークンが見つからない場合）
 */
function getBearerToken(req: Request): string | null {
  // Authorizationヘッダーを取得（大文字小文字を区別しない）
  const header =
    req.headers.authorization ||
    (req.headers as Record<string, string | string[] | undefined>).Authorization;
  if (!header) return null;
  
  // ヘッダー値が配列の場合は最初の要素を使用
  const value = Array.isArray(header) ? header[0] : header;
  const [type, token] = value.split(" ");
  
  // トークンタイプが"bearer"で、トークンが存在する場合のみ返す
  if (type?.toLowerCase() !== "bearer" || !token) return null;
  return token;
}

/**
 * リクエストを認証する（AWS Cognito）
 * 
 * リクエストに含まれるBearerトークンを使用して、AWS Cognitoで認証を行う。
 * JWTを検証し、ユーザー情報をデータベースに保存/更新してからユーザーオブジェクトを返す。
 * 
 * 認証フロー:
 * 1. リクエストからBearerトークンを取得
 * 2. Cognitoの設定（JWKS、Client ID、Issuer）を確認
 * 3. JWTを検証（署名、発行者、有効期限、audience）
 * 4. トークンのペイロードからユーザー情報（openId、name、email）を取得
 * 5. データベースからユーザーを取得、存在しない場合は作成
 * 6. ユーザー情報を更新（最後のログイン時刻を更新）
 * 7. ユーザーオブジェクトを返す
 * 
 * セキュリティ:
 * - JWTの署名を検証して、トークンの改ざんを検知
 * - 発行者（issuer）とaudience（client ID）を検証
 * - 有効期限を検証（joseライブラリが自動的に行う）
 * 
 * @param req - Expressリクエストオブジェクト
 * @returns 認証されたユーザーオブジェクト
 * @throws {ForbiddenError} トークンが存在しない、設定が不完全、またはトークンの検証に失敗した場合
 * 
 * @example
 * const user = await authenticateRequest(req);
 * console.log(`認証されたユーザー: ${user.name}`);
 */
export async function authenticateRequest(
  req: Request,
  options: { updateUser?: boolean } = {}
): Promise<User> {
  // Bearerトークンを取得
  const token = getBearerToken(req);
  if (!token) {
    throw ForbiddenError("Missing Authorization token");
  }
  
  // Cognitoの設定を確認
  // JWKSとClient IDが設定されていない場合は認証できない
  if (!jwks || !ENV.cognitoClientId) {
    throw ForbiddenError("Cognito auth is not configured");
  }

  // 発行者URLを取得
  const issuer = getIssuer();
  if (!issuer) {
    throw ForbiddenError("Cognito issuer is not configured");
  }

  // JWTを検証
  // 署名、発行者、audience、有効期限を検証する
  const { payload } = await jwtVerify(token, jwks, {
    issuer,
    audience: ENV.cognitoClientId,
  });

  // トークンタイプを確認（IDトークンのみ許可）
  // token_useが"id"以外の場合は拒否
  if (payload.token_use && payload.token_use !== "id") {
    throw ForbiddenError("Invalid token type");
  }

  // ペイロードからユーザー情報を取得
  // sub（subject）はユーザーの一意識別子（openIdとして使用）
  const openId = typeof payload.sub === "string" ? payload.sub : null;
  if (!openId) {
    throw ForbiddenError("Invalid token payload");
  }

  // デバッグ: ペイロードの内容を確認
  if (process.env.DEBUG_AUTH === "true") {
    console.log("[Cognito] Token payload keys:", Object.keys(payload));
    console.log("[Cognito] Token payload (name-related):", {
      name: payload.name ? "[REDACTED]" : undefined,
      given_name: payload.given_name ? "[REDACTED]" : undefined,
      family_name: payload.family_name ? "[REDACTED]" : undefined,
      "cognito:username": payload["cognito:username"] ? "[REDACTED]" : undefined,
      email: payload.email ? "[REDACTED]" : undefined,
    });
  }

  // オプションのユーザー情報を取得
  // CognitoのIDトークンには、name、given_name、family_name、cognito:usernameなどが含まれる可能性がある
  // 優先順位: name > given_name + family_name > cognito:username
  let name: string | null = null;
  if (typeof payload.name === "string" && payload.name) {
    name = payload.name;
  } else if (typeof payload.given_name === "string" || typeof payload.family_name === "string") {
    // given_nameとfamily_nameを組み合わせる
    const givenName = typeof payload.given_name === "string" ? payload.given_name : "";
    const familyName = typeof payload.family_name === "string" ? payload.family_name : "";
    name = `${givenName} ${familyName}`.trim() || null;
  } else if (typeof payload["cognito:username"] === "string" && payload["cognito:username"]) {
    // フォールバック: cognito:usernameを使用
    name = payload["cognito:username"];
  }
  
  const email = typeof payload.email === "string" ? payload.email : null;
  
  // デバッグ: 抽出されたユーザー情報を確認（個人情報はマスク）
  if (process.env.DEBUG_AUTH === "true") {
    console.log("[Cognito] Extracted user info:", {
      openId: openId ? `${openId.substring(0, 10)}...` : null,
      name: name ? "[REDACTED]" : null,
      email: email ? "[REDACTED]" : null,
    });
  }

  const shouldUpdateUser = options.updateUser === true;

  // データベースからユーザーを取得
  let user = await db.getUserByOpenId(openId);
  if (!user) {
    // ユーザーが存在しない場合は新規作成
    await db.upsertUser({
      openId,
      name,
      email,
      loginMethod: "cognito",
      lastSignedIn: new Date(),
    });
    user = await db.getUserByOpenId(openId);
  } else if (shouldUpdateUser && !user.deletedAt) {
    await db.upsertUser({
      openId,
      name: name || user.name || null,
      email: email || user.email || null,
      loginMethod: user.loginMethod || "cognito",
      lastSignedIn: new Date(),
    });
    user = await db.getUserByOpenId(openId);
  }

  // ユーザーが取得できない場合（データベースエラーなど）
  if (!user) {
    throw ForbiddenError("User not found");
  }

  return user;
}

File: server_python/app/core/cognito.py
"""
AWS Cognito authentication
"""
import json
from typing import Optional
from fastapi import Request, HTTPException, status
from jose import jwt, jwk
from jose.exceptions import JWTError, ExpiredSignatureError
import httpx
from app.core.config import settings
from app.core.database import get_user_by_open_id, upsert_user


def get_bearer_token(request: Request) -> Optional[str]:
    """Extract Bearer token from Authorization header"""
    auth_header = request.headers.get("authorization") or request.headers.get("Authorization")
    if not auth_header:
        return None
    
    parts = auth_header.split(" ", 1)
    if len(parts) != 2 or parts[0].lower() != "bearer":
        return None
    
    return parts[1]


async def get_jwks() -> dict:
    """Get JWKS from Cognito"""
    jwks_url = settings.cognito_jwks_url_final
    if not jwks_url:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Cognito JWKS URL is not configured"
        )
    
    async with httpx.AsyncClient() as client:
        response = await client.get(jwks_url)
        response.raise_for_status()
        return response.json()


async def verify_cognito_token(token: str, access_token: Optional[str] = None) -> dict:
    """Verify Cognito JWT token"""
    if not settings.cognito_client_id:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Cognito client ID is not configured"
        )
    
    issuer = settings.cognito_issuer
    if not issuer:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Cognito issuer is not configured"
        )
    
    # Get JWKS
    jwks_data = await get_jwks()
    
    # Decode token header to get key ID
    try:
        unverified_header = jwt.get_unverified_header(token)
    except jwt.PyJWTError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid token header"
        )
    kid = unverified_header.get("kid")
    
    if not kid:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token missing key ID"
        )
    
    # Find the key in JWKS
    key = None
    for jwk_key in jwks_data.get("keys", []):
        if jwk_key.get("kid") == kid:
            key = jwk_key
            break
    
    if not key:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Key not found in JWKS"
        )
    
    # Construct the public key
    public_key = jwk.construct(key)
    
    # Verify and decode the token
    try:
        options = {"verify_at_hash": bool(access_token)}
        payload = jwt.decode(
            token,
            public_key,
            algorithms=["RS256"],
            audience=settings.cognito_client_id,
            issuer=issuer,
            options=options,
            access_token=access_token,
        )
    except ExpiredSignatureError:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Token has expired"
        )
    except JWTError as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail=f"Token validation failed: {str(e)}"
        )
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail=f"Token verification failed: {str(e)}"
        )
    
    # Check token type (only ID tokens are allowed)
    if payload.get("token_use") and payload.get("token_use") != "id":
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid token type"
        )
    
    return payload


async def authenticate_request(
    request,
    update_user: bool = False,
    access_token: Optional[str] = None
):
    """Authenticate request using Cognito Bearer token"""
    # Support both Request object and dict with headers
    if hasattr(request, 'headers'):
        token = get_bearer_token(request)
    elif isinstance(request, dict) and 'headers' in request:
        # Mock request object
        class MockRequest:
            def __init__(self, headers):
                self.headers = headers
        mock_req = MockRequest(request['headers'])
        token = get_bearer_token(mock_req)
    else:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Invalid request object"
        )
    
    if not token:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Missing Authorization token"
        )
    
    # Verify token
    payload = await verify_cognito_token(token, access_token=access_token)
    
    # Extract user information
    open_id = payload.get("sub")
    if not open_id or not isinstance(open_id, str):
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail="Invalid token payload"
        )
    
    # Extract name (priority: name > given_name + family_name > cognito:username)
    name = None
    if payload.get("name") and isinstance(payload.get("name"), str):
        name = payload["name"]
    elif payload.get("given_name") or payload.get("family_name"):
        given_name = payload.get("given_name", "") or ""
        family_name = payload.get("family_name", "") or ""
        name = f"{given_name} {family_name}".strip() or None
    elif payload.get("cognito:username"):
        name = payload["cognito:username"]
    
    email = payload.get("email")
    if email and not isinstance(email, str):
        email = None
    
    # Get or create user
    user = await get_user_by_open_id(open_id)
    if not user:
        # Create new user
        await upsert_user({
            "openId": open_id,
            "name": name,
            "email": email,
            "loginMethod": "cognito",
        })
        user = await get_user_by_open_id(open_id)
    elif update_user and not user.get("deletedAt"):
        # Update existing user
        await upsert_user({
            "openId": open_id,
            "name": name or user.get("name"),
            "email": email or user.get("email"),
            "loginMethod": user.get("loginMethod") or "cognito",
        })
        user = await get_user_by_open_id(open_id)
    
    if not user:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="User not found"
        )
    
    return user
